### DISCUSSION TITLE
Async WIFI and MQTT connections?

### DISCUSSION THREAD

**Discussion by mikep1998** [arendst/Tasmota#12122] [Ideas]
Hello, has anyone looked into using an async wifi and mqtt classes?

I'm currently using the Display code and it is working very well. But when tasmota is connecting to WIFI or MQTT or reconnecting the screen will hang for a few seconds until the connect times out or the connection is made.

Thanks
Mike

**Comment by jeroenst** [comment]
I wonder if communications also affect the responsiveness of pushbuttons for dimmers and light switches...

**Reply by mikep1998** [reply]
Yes everything is frozen while attempting to connect.

**Reply by jeroenst** [reply]
Hmmz, that is not good at all...

**Reply by barbudor** [reply]
This is also the case for Ntp and Tasmota will froze while attempting to connect to a NTP server. If not responding the peformance is seriously hit.

I think the probem is that many libraries available for Arduino are not asynchronous. Normally on a LAN it should not be a problem and MQTT connection should occur within less than a second. I even not noticed delays when using cloud based MQTT servers like AdafruitIO or Flespi.
Now, when connection problem arise, it may definitively impact Tasmota and responsiveness.

**Reply by jeroenst** [reply]
That's why i use https://github.com/me-no-dev/ESPAsyncTCP and https://github.com/marvinroger/async-mqtt-client in my own code.

**Reply by mikep1998** [reply]
Yes, I have used ESPAsyncTCP too. Has anyone tried to change the current WIFI to an Async WIFI? If so where there any issues?

**Reply by s-hadinger** [reply]
There is a first freeze when Wifi is connecting to the AP, I doubt we can do much here. The second freeze when connecting MQTT could indeed benefit from async TCP, however it is inside the Arduino lib, I don't know how easy/hard it is to change.

For ESP32, there is now a lightweight MQTT client at FreeRTOS level, which is async by design. That could be something to look at in near future (but not for ESP8266)

**Reply by mikep1998** [reply]
I used WiFiMulti and the connect to AP is asynchronous too.  So connecting and reconnect wifi is not blocked. AsyncMqttClient also works well for not blocking.

The other thing that may also work is moving the LVGL process back to its own processor. That way Tasmota can wait for the connection while LVGL continues to respond to touch  and screen updates in its own processor.
What issue were observed when LVGL was in its own processor?

**Reply by s-hadinger** [reply]
I don't know about WifiMulti... We'll need to explore this.

For LVGL, it was not running on its own processor but on its own thread. This meant that animations were continuing even during blocking events like OTA upgrade. This sounds nice, BUT...
It also means that Tasmota and LVGL run in 2 different threads and that any interactions between them needs to be protected. All calls to LVGL must be gated by a Mutex (medium impact, doable). However this completely kills the ability to call Berry callbacks from LVGL code. I plan to be able to create LVGL widgets in Berry, and this relies on synchronous callbacks. I don't see any solution except having a second Berry VM running in LVGL thread.

Honestly the benefit/cost ratio is too high for me, just to have smooth animations in rare situations. I will keep LVGL in Tasmota thread for now.

Unless you have a solution :)

**Reply by mikep1998** [reply]
I agree LVGL is its own thread would be to much work for the benefits.

Async wifi and mqtt are the way to go.
